pkg/prometheus/operator.go

```diff
@@ -247,3 +264,59 @@ func (sr *StatusReporter) Process(ctx context.Context, p monitoringv1.Prometheus
 
 	return &pStatus, nil
 }
+
+// AddStatus add the latest status in serviceMonitor, podMonitor, probes and scrapeConfig
+// resources selected by the Prometheus or PrometheusAgent.
+func (ru *ConfigResourceSyncer[T]) AddStatus(ctx context.Context, p metav1.Object, resources ResourcesSelection[T]) error {
+	for key, res := range resources {
+		condition := monitoringv1.ConfigResourceCondition{
+			Type:               monitoringv1.Accepted,
+			Status:             monitoringv1.ConditionTrue,
+			LastTransitionTime: metav1.Now(),
+			Reason:             res.reason,
+		}
+
+		if res.err != nil {
+			condition.Status = monitoringv1.ConditionFalse
+			condition.Message = res.err.Error()
+		}
+
+		binding := monitoringv1.WorkloadBinding{
+			Namespace:  p.GetNamespace(),
+			Name:       p.GetName(),
+			Resource:   ru.gvr.Resource,
+			Group:      ru.gvr.Group,
+			Conditions: []monitoringv1.ConfigResourceCondition{condition},
+		}
+		switch r := any(res.resource).(type) {
```

Comment:
It feels a bit awkward to have a generic type and then fallback to type-casting.